### ISSUE TITLE
render a 3d mesh against a defined background image

### ISSUE THREAD

**Issue by fairydora** [facebookresearch/pytorch3d#577]
Hello,
Is it possible to render a 3d mesh against a defined background image?
thanks for any help :)

**Comment by nikhilaravi** [comment]
Hi @fairydora you could define your own blending function and use this in a custom shader for this.  For example here is an code example (it has not been tested but should give you an idea of what to do) : 

```
class BlendParams(NamedTuple):
    sigma: float = 1e-4
    gamma: float = 1e-4
    background_color: Sequence = (1.0, 1.0, 1.0)
    # add background image as a param here 
    background_image: # (H, W, 3) 

def hard_rgb_blend_with_background(colors, fragments, blend_params) -> torch.Tensor:
    N, H, W, K = fragments.pix_to_face.shape
    device = fragments.pix_to_face.device

    # Mask for the background.
    is_background = fragments.pix_to_face[..., 0] < 0  # (N, H, W)

    if torch.is_tensor(blend_params.background_image):
        background_image = blend_params.background_image.to(device)
    else:
        background_image = colors.new_tensor(blend_params.background_image)  # (H, W, 3)

    # Format background image
    # (H, W, 3) -> (N, H, W, 3) -> select only pixels which are in the background using the mask 
    background_image_masked = background_image[None, ...].expand(N, -1, -1, -1)[is_background[..., None]] 

    # Set background color pixels to the colors from the background image
    pixel_colors = colors[..., 0, :].masked_scatter(
        is_background[..., None],
        background_image_masked
    )  # (N, H, W, 3)

    # Concat with the alpha channel.
    alpha = torch.ones((N, H, W, 1), dtype=colors.dtype, device=device)
    return torch.cat([pixel_colors, alpha], dim=-1)  # (N, H, W, 4)
```

You can then create your own shader class which replaces the shading function with your custom function above e.g. replace this line:

https://github.com/facebookresearch/pytorch3d/blob/4bb3fff52b7e26ec0f013021cb26fab7db3d8e0b/pytorch3d/renderer/mesh/shader.py#L79

**NOTE:** This simple example assumes the image size of the background image is the same as the rendered image and also uses the same image for all examples in the batch. You can customize this as required.

**Comment by fairydora** [comment]
Thank you @nikhilaravi :)
It's really kind of you to reply so fast! 
I will now put this solution to good use. 
I really appreciate your help :)

**Comment by nikhilaravi** [comment]
@fairydora did you try it out? Do you still need help with this question?

**Comment by fairydora** [comment]
@nikhilaravi  
Hi Nikhila, 
Thank you for your interest :)
I've done something similar but simpler, but got inspired by your code; 
I've got the alpha from the silhouette renderer, but cleaned it a bit to contain only 0 and 1, then composited the rendered image with alpha on top of the background, like below:

#render = rendered mesh

inverted_mask = abs(mask-1)
bckg = torch.tensor(self.background_image / 255, dtype=torch.float64, device=self.device)
channels = [bckg[..., i] * inverted_mask + render[..., i] * mask for i in range(3)]
composited = torch.stack(channels, axis=2)

Thank you very much for your help! :)

**Comment by nikhilaravi** [comment]
Great, glad you were able to resolve your issue!
